plumbing/transport/fetch_pack.go

```diff
@@ -128,6 +128,28 @@ func (r *UploadPackRequest) Have(h ...plumbing.Hash) {
 	r.Haves = append(r.Haves, h...)
 }
 
+func (r *UploadPackRequest) IsEmpty() bool {
+	if len(r.Haves) < len(r.Wants) {
+		return false
+	}
+
+	for _, h := range r.Wants {
+		found := false
+		for _, oh := range r.Haves {
+			if h == oh {
+				found = true
+				break
+			}
+		}
```

Comment:
Given the use we are doing of `UploadPackRequest.Wants` and  `UploadPackRequest.Haves` we should turn them into maps (actually, sets).  Also change the name of `h` or `w` and `oh` to `have`.

This two nested for loops can be clarified a lot: what we are doing is checking if `r.Wants` is a subset of `r.Haves`. So I would substitute the loop for:

```
if areSetSubset(r.Haves, r.Wants) {
    return true
}
```

and create the function `areSetSubset` somewhere.

We can also define a new type for a list of references (wants and haves) and define some common set operations like `complementary set` and `is subset`.

Also, I think `r.Wants` and `r.Haves` should be private, unless I am missing something.